File: src-tauri/src/kanban.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ═══════════════════════════════════════════
// Types — mirrors Hermes Studio kanban API
// ═══════════════════════════════════════════

pub type KanbanTaskStatus = String; // triage|todo|scheduled|ready|running|blocked|review|done|archived

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KanbanTask {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub body: Option<String>,
    #[serde(default)]
    pub assignee: Option<String>,
    pub status: KanbanTaskStatus,
    #[serde(default)]
    pub priority: i32,
    #[serde(default)]
    pub created_by: Option<String>,
    #[serde(default)]
    pub created_at: f64,
    #[serde(default)]
    pub started_at: Option<f64>,
    #[serde(default)]
    pub completed_at: Option<f64>,
    #[serde(default)]
    pub workspace_kind: String,
    #[serde(default)]
    pub workspace_path: Option<String>,
    #[serde(default)]
    pub tenant: Option<String>,
    #[serde(default)]
    pub result: Option<String>,
    #[serde(default)]
    pub skills: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KanbanBoard {
    pub slug: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub icon: String,
    #[serde(default)]
    pub color: String,
    #[serde(default)]
    pub created_at: Option<f64>,
    #[serde(default)]
    pub archived: bool,
    #[serde(default)]
    pub counts: HashMap<String, u32>,
    #[serde(default)]
    pub total: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KanbanStats {
    pub by_status: HashMap<String, u32>,
    pub by_assignee: HashMap<String, u32>,
    pub total: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KanbanAssignee {
    pub name: String,
    pub on_disk: bool,
    #[serde(default)]
    pub counts: Option<HashMap<String, u32>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct KanbanComment {
    pub id: i64,
    pub task_id: String,
    pub author: String,
    pub body: String,
    pub created_at: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct KanbanEvent {
    pub id: i64,
    pub task_id: String,
    pub kind: String,
    #[serde(default)]
    pub payload: Option<HashMap<String, serde_json::Value>>,
    pub created_at: f64,
    #[serde(default)]
    pub run_id: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct KanbanRun {
    pub id: i64,
    pub task_id: String,
    #[serde(default)]
    pub profile: Option<String>,
    pub status: String,
    pub started_at: f64,
    #[serde(default)]
    pub ended_at: Option<f64>,
    #[serde(default)]
    pub outcome: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct KanbanTaskDetail {
    pub task: KanbanTask,
    #[serde(default)]
    pub comments: Vec<KanbanComment>,
    #[serde(default)]
    pub events: Vec<KanbanEvent>,
    #[serde(default)]
    pub runs: Vec<KanbanRun>,
}

// ═══════════════════════════════════════════
// Hermes Studio API helpers
// ═══════════════════════════════════════════

fn hermes_base_url() -> String {
    std::env::var("HERMES_STUDIO_URL")
        .unwrap_or_else(|_| "http://localhost:8648".to_string())
}

fn hermes_client() -> reqwest::blocking::Client {
    reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .unwrap_or_default()
}

// ═══════════════════════════════════════════
// Tauri Commands
// ═══════════════════════════════════════════

#[tauri::command]
pub async fn kanban_list_boards(include_archived: Option<bool>) -> Result<Vec<KanbanBoard>, String> {
    let mut url = format!("{}/api/hermes/kanban/boards", hermes_base_url());
    if include_archived.unwrap_or(false) {
        url.push_str("?includeArchived=true");
    }
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    let boards: Vec<KanbanBoard> = serde_json::from_value(
        raw.get("boards").cloned().unwrap_or(serde_json::Value::Array(vec![])),
    )
    .map_err(|e| e.to_string())?;
    Ok(boards)
}

#[tauri::command]
pub async fn kanban_list_tasks(
    board: Option<String>,
    status: Option<String>,
    assignee: Option<String>,
    include_archived: Option<bool>,
) -> Result<Vec<KanbanTask>, String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let mut url = format!("{}/api/hermes/kanban?board={}", hermes_base_url(), board_slug);
    if let Some(s) = &status {
        url.push_str(&format!("&status={}", s));
    }
    if let Some(a) = &assignee {
        url.push_str(&format!("&assignee={}", a));
    }
    if include_archived.unwrap_or(false) {
        url.push_str("&includeArchived=true");
    }
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    let tasks: Vec<KanbanTask> = serde_json::from_value(
        raw.get("tasks").cloned().unwrap_or(serde_json::Value::Array(vec![])),
    )
    .map_err(|e| e.to_string())?;
    Ok(tasks)
}

#[tauri::command]
pub async fn kanban_get_task(task_id: String, board: Option<String>) -> Result<KanbanTaskDetail, String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/{}?board={}",
        hermes_base_url(),
        task_id,
        board_slug
    );
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let detail: KanbanTaskDetail = resp.json().map_err(|e| e.to_string())?;
    Ok(detail)
}

#[tauri::command]
pub async fn kanban_get_stats(board: Option<String>) -> Result<KanbanStats, String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/stats?board={}",
        hermes_base_url(),
        board_slug
    );
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    let stats: KanbanStats = serde_json::from_value(
        raw.get("stats").cloned().unwrap_or(serde_json::Value::Null),
    )
    .map_err(|e| e.to_string())?;
    Ok(stats)
}

#[tauri::command]
pub async fn kanban_get_assignees(board: Option<String>) -> Result<Vec<KanbanAssignee>, String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/assignees?board={}",
        hermes_base_url(),
        board_slug
    );
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    let assignees: Vec<KanbanAssignee> = serde_json::from_value(
        raw.get("assignees")
            .cloned()
            .unwrap_or(serde_json::Value::Array(vec![])),
    )
    .map_err(|e| e.to_string())?;
    Ok(assignees)
}

#[derive(Debug, Deserialize)]
pub struct KanbanCreateRequest {
    pub title: String,
    #[serde(default)]
    pub body: Option<String>,
    #[serde(default)]
    pub assignee: Option<String>,
    #[serde(default)]
    pub priority: Option<i32>,
}

#[tauri::command]
pub async fn kanban_create_task(
    data: KanbanCreateRequest,
    board: Option<String>,
) -> Result<KanbanTask, String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!("{}/api/hermes/kanban?board={}", hermes_base_url(), board_slug);
    let client = hermes_client();
    let mut body = serde_json::json!({ "title": data.title });
    if let Some(b) = &data.body {
        body["body"] = serde_json::Value::String(b.clone());
    }
    if let Some(a) = &data.assignee {
        body["assignee"] = serde_json::Value::String(a.clone());
    }
    if let Some(p) = data.priority {
        body["priority"] = serde_json::Value::Number(p.into());
    }
    let resp = client.post(&url).json(&body).send().map_err(|e| e.to_string())?;
    let raw: serde_json::Value = resp.json().map_err(|e| e.to_string())?;
    let task: KanbanTask = serde_json::from_value(
        raw.get("task").cloned().unwrap_or(serde_json::Value::Null),
    )
    .map_err(|e| e.to_string())?;
    Ok(task)
}

#[tauri::command]
pub async fn kanban_complete_tasks(
    task_ids: Vec<String>,
    summary: Option<String>,
    board: Option<String>,
) -> Result<(), String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/complete?board={}",
        hermes_base_url(),
        board_slug
    );
    let mut body = serde_json::json!({ "task_ids": task_ids });
    if let Some(s) = &summary {
        body["summary"] = serde_json::Value::String(s.clone());
    }
    let client = hermes_client();
    client.post(&url).json(&body).send().map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn kanban_block_task(
    task_id: String,
    reason: String,
    board: Option<String>,
) -> Result<(), String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/{}/block?board={}",
        hermes_base_url(),
        task_id,
        board_slug
    );
    let client = hermes_client();
    client
        .post(&url)
        .json(&serde_json::json!({ "reason": reason }))
        .send()
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn kanban_unblock_tasks(task_ids: Vec<String>, board: Option<String>) -> Result<(), String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/unblock?board={}",
        hermes_base_url(),
        board_slug
    );
    let client = hermes_client();
    client
        .post(&url)
        .json(&serde_json::json!({ "task_ids": task_ids }))
        .send()
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn kanban_assign_task(
    task_id: String,
    profile: String,
    board: Option<String>,
) -> Result<(), String> {
    let board_slug = board.unwrap_or_else(|| "default".to_string());
    let url = format!(
        "{}/api/hermes/kanban/{}/assign?board={}",
        hermes_base_url(),
        task_id,
        board_slug
    );
    let client = hermes_client();
    client
        .post(&url)
        .json(&serde_json::json!({ "profile": profile }))
        .send()
        .map_err(|e| e.to_string())?;
    Ok(())
}

File: src/stores/kanban.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { invoke } from "@tauri-apps/api/core";

export type KanbanTaskStatus =
  | "triage"
  | "todo"
  | "scheduled"
  | "ready"
  | "running"
  | "blocked"
  | "review"
  | "done"
  | "archived";

export interface KanbanTask {
  id: string;
  title: string;
  body: string | null;
  assignee: string | null;
  status: KanbanTaskStatus;
  priority: number;
  created_by: string | null;
  created_at: number;
  started_at: number | null;
  completed_at: number | null;
  workspace_kind: string;
  workspace_path: string | null;
  tenant: string | null;
  result: string | null;
  skills: string[] | null;
}

export interface KanbanBoard {
  slug: string;
  name: string;
  description: string;
  icon: string;
  color: string;
  created_at: number | null;
  archived: boolean;
  counts: Record<string, number>;
  total: number;
}

export interface KanbanStats {
  by_status: Record<string, number>;
  by_assignee: Record<string, number>;
  total: number;
}

export interface KanbanAssignee {
  name: string;
  on_disk: boolean;
  counts: Record<string, number> | null;
}

export const KANBAN_COLUMNS: KanbanTaskStatus[] = [
  "triage",
  "todo",
  "ready",
  "running",
  "blocked",
  "review",
  "done",
];

export const KANBAN_STATUS_LABELS: Record<string, string> = {
  triage: "分类",
  todo: "待办",
  scheduled: "计划",
  ready: "就绪",
  running: "执行中",
  blocked: "阻塞",
  review: "审查",
  done: "完成",
  archived: "归档",
};

export const KANBAN_STATUS_ICONS: Record<string, string> = {
  triage: "circle",
  todo: "○",
  scheduled: "loader",
  ready: "eye",
  running: "check",
  blocked: "ban",
  review: "archive",
  done: "✓",
  archived: "▪",
};

export const KANBAN_STATUS_COLORS: Record<string, string> = {
  triage: "#94a3b8",
  todo: "#38bdf8",
  scheduled: "#06b6d4",
  ready: "#f59e0b",
  running: "#8b5cf6",
  blocked: "#ef4444",
  review: "#ec4899",
  done: "#22c55e",
  archived: "#64748b",
};

export const useKanbanStore = defineStore("kanban", () => {
  const tasks = ref<KanbanTask[]>([]);
  const boards = ref<KanbanBoard[]>([]);
  const stats = ref<KanbanStats | null>(null);
  const assignees = ref<KanbanAssignee[]>([]);
  const selectedBoard = ref("default");
  const loading = ref(false);
  const filterStatus = ref<string | null>(null);
  const filterAssignee = ref<string | null>(null);
  const selectedTaskId = ref<string | null>(null);

  const tasksByStatus = computed(() => {
    const map: Record<string, KanbanTask[]> = {};
    for (const col of KANBAN_COLUMNS) {
      map[col] = [];
    }
    for (const task of tasks.value) {
      if (task.status === "archived") continue;
      if (!map[task.status]) map[task.status] = [];
      map[task.status].push(task);
    }
    // Sort by priority (high first), then created_at (newest first)
    for (const col of KANBAN_COLUMNS) {
      map[col].sort((a, b) => {
        if (b.priority !== a.priority) return b.priority - a.priority;
        return b.created_at - a.created_at;
      });
    }
    return map;
  });

  const activeBoards = computed(() => {
    return boards.value.filter((b) => !b.archived);
  });

  async function fetchBoards(includeArchived = false) {
    try {
      const result: KanbanBoard[] = await invoke("kanban_list_boards", {
        includeArchived,
      });
      boards.value = result;
    } catch (e) {
      console.error("Failed to fetch boards:", e);
    }
  }

  async function fetchTasks() {
    loading.value = true;
    try {
      const result: KanbanTask[] = await invoke("kanban_list_tasks", {
        board: selectedBoard.value,
        status: filterStatus.value || undefined,
        assignee: filterAssignee.value || undefined,
        includeArchived: true,
      });
      tasks.value = result;
    } catch (e) {
      console.error("Failed to fetch tasks:", e);
    } finally {
      loading.value = false;
    }
  }

  async function fetchStats() {
    try {
      const result: KanbanStats = await invoke("kanban_get_stats", {
        board: selectedBoard.value,
      });
      stats.value = result;
    } catch (e) {
      console.error("Failed to fetch stats:", e);
    }
  }

  async function fetchAssignees() {
    try {
      const result: KanbanAssignee[] = await invoke("kanban_get_assignees", {
        board: selectedBoard.value,
      });
      assignees.value = result;
    } catch (e) {
      console.error("Failed to fetch assignees:", e);
    }
  }

  async function createTask(data: {
    title: string;
    body?: string;
    assignee?: string;
    priority?: number;
  }) {
    try {
      const task: KanbanTask = await invoke("kanban_create_task", {
        data,
        board: selectedBoard.value,
      });
      tasks.value.unshift(task);
      await Promise.all([fetchStats(), fetchBoards()]);
      return task;
    } catch (e) { console.error('[Kanban] createTask failed:', e); }
  }

  async function completeTasks(taskIds: string[], summary?: string) {
    try {
      await invoke("kanban_complete_tasks", {
        taskIds,
        summary,
        board: selectedBoard.value,
      });
      for (const id of taskIds) {
        const task = tasks.value.find((t) => t.id === id);
        if (task) task.status = "done";
      }
      await Promise.all([fetchStats(), fetchBoards()]);
    } catch (e) { console.error('[Kanban] completeTasks failed:', e); }
  }

  async function blockTask(taskId: string, reason: string) {
    try {
      await invoke("kanban_block_task", {
        taskId,
        reason,
        board: selectedBoard.value,
      });
      const task = tasks.value.find((t) => t.id === taskId);
      if (task) task.status = "blocked";
      await Promise.all([fetchStats(), fetchBoards()]);
    } catch (e) { console.error('[Kanban] blockTask failed:', e); }
  }

  async function unblockTasks(taskIds: string[]) {
    try {
      await invoke("kanban_unblock_tasks", {
        taskIds,
        board: selectedBoard.value,
      });
      for (const id of taskIds) {
        const task = tasks.value.find((t) => t.id === id);
        if (task) task.status = "ready";
      }
      await Promise.all([fetchStats(), fetchBoards()]);
    } catch (e) { console.error('[Kanban] unblockTasks failed:', e); }
  }

  async function assignTask(taskId: string, profile: string) {
    try {
      await invoke("kanban_assign_task", {
        taskId,
        profile,
        board: selectedBoard.value,
      });
      const task = tasks.value.find((t) => t.id === taskId);
      if (task) task.assignee = profile;
      await Promise.all([fetchStats(), fetchAssignees()]);
    } catch (e) { console.error('[Kanban] assignTask failed:', e); }
  }

  async function refreshAll() {
    await Promise.all([fetchBoards(), fetchTasks(), fetchStats(), fetchAssignees()]);
  }

  function setBoard(board: string) {
    selectedBoard.value = board;
    selectedTaskId.value = null;
  }

  function setFilter(key: "status" | "assignee", value: string | null) {
    if (key === "status") filterStatus.value = value;
    else filterAssignee.value = value;
  }

  function timeAgo(ts: number): string {
    const diff = Date.now() / 1000 - ts;
    if (diff < 60) return "刚刚";
    if (diff < 3600) return `${Math.floor(diff / 60)} 分钟前`;
    if (diff < 86400) return `${Math.floor(diff / 3600)} 小时前`;
    if (diff < 45 * 86400) return `${Math.floor(diff / 86400)} 天前`;
    return `${Math.floor(diff / (30 * 86400))} 月前`;
  }

  return {
    tasks,
    boards,
    stats,
    assignees,
    selectedBoard,
    loading,
    filterStatus,
    filterAssignee,
    selectedTaskId,
    tasksByStatus,
    activeBoards,
    fetchBoards,
    fetchTasks,
    fetchStats,
    fetchAssignees,
    createTask,
    completeTasks,
    blockTask,
    unblockTasks,
    assignTask,
    refreshAll,
    setBoard,
    setFilter,
    timeAgo,
  };
});
